Improve performance of WebUtility

hughbe opened the pull request:
## Summary
No escaping needed:
- x5 performance improvement
- 0 allocations

Basic escaping needed (single char):
- 20% performance improvement
- Allocations cut by a third

Advanced escaping needed (surrogate pair):
- 15% performance improvement
- Allocations cut by a third

## Benchmark Results
- No breaking change in behaviour in these validation tests
- Times measured in seconds

![benchmark](https://cloud.githubusercontent.com/assets/1275900/14318468/e4ba70d0-fc04-11e5-8a42-3604666dbab3.png)

## Benchmark Code and !!Validation Tests!!
- **To ensure there were no breaking changes** I ran the new output against the old output for every single char and surrogate pair in the UTF16 plane.
```c#
static unsafe void Main(string[] args)
{
    PerformanceTests();
    VerificationTests();    
    Console.ReadLine();
}

private static void PerformanceTests()
{
    Console.WriteLine("**** MEASURE NO ESCAPING ****");
    TimeAction("Old: ", () => Old.UrlEncode("abcxyz"));
    TimeAction("New: ", () => New.UrlEncode("abcxyz"));

    Console.WriteLine("**** MEASURE SPACES ****");
    TimeAction("Old: ", () => Old.UrlEncode("abc xyz"));
    TimeAction("New: ", () => New.UrlEncode("abc xyz"));

    Console.WriteLine("**** MEASURE ESCAPING 1 TIMES ****");
    TimeAction("Old: ", () => Old.UrlEncode("\u007F"));
    TimeAction("New: ", () => New.UrlEncode("\u007F"));

    Console.WriteLine("**** MEASURE ESCAPING 2 TIMES ****");
    TimeAction("Old: ", () => Old.UrlEncode("\u07FF"));
    TimeAction("New: ", () => New.UrlEncode("\u07FF"));

    Console.WriteLine("**** MEASURE ESCAPING 3 TIMES ****");
    TimeAction("Old: ", () => Old.UrlEncode("\u1234"));
    TimeAction("New: ", () => New.UrlEncode("\u1234"));

    Console.WriteLine("**** MEASURE ESCAPING SURROGATE PAIR ****");
    TimeAction("Old: ", () => Old.UrlEncode("\uD801\uDC37"));
    TimeAction("New: ", () => New.UrlEncode("\uD801\uDC37"));
}

public static void TimeAction(string prefix, Action action)
{
    action();
    GC.Collect();

    var sw = new Stopwatch();
    for (int iter = 0; iter < 5; iter++)
    {
        int gen0 = GC.CollectionCount(0);
        sw.Restart();
        for (int i = 0; i < 100000000; i++)
        {
            action();
        }
        sw.Stop();
        Console.WriteLine($"{prefix}Time: {sw.Elapsed.TotalSeconds}\tGC0: {GC.CollectionCount(0) - gen0}");
    }
}

private static void VerificationTests()
{
    for (int i = 0; i < char.MaxValue; i++)
    {
        char c = (char)i;

        string oldString = Old.UrlEncode(c.ToString());
        string newString = New.UrlEncode(c.ToString());
        if (!oldString.Equals(newString))
        {
            Console.WriteLine($"FAIL for char {i}. Expected: {oldString}. Actual: {newString}");
        }
    }
    for (int i = 0x10000; i < 0x10FFFF; i++)
    {
        string utf32 = char.ConvertFromUtf32(i);
        string oldString = Old.UrlEncode(utf32);
        string newString = New.UrlEncode(utf32);
        if (!oldString.Equals(newString))
        {
            Console.WriteLine($"FAIL for char {i}. Expected: {oldString}. Actual: {newString}");
        }
    }
    Console.WriteLine("Passes Validation Tests");
}
```

/cc @davidsh, @stephentoub, @jamesqo

hughbe commented:
I was just thinking, we don't have many good validation tests for WebUtility. Would there be a value in adding tests for all unicode code points, matched to a database of values. Of course, this would add something like a 25K line file to the project, but its been done before (e.g. Unicode.8.0.0.txt data)

stephentoub commented:
I don't understand how this could improve performance: you added dead code that's not actually called from anywhere.

hughbe commented:
Oh crumbs I forgot to actually add the code! Forgot to copy it over from visual studio test prince

hughbe commented:
Temporarily closing this - I found a failing test

hughbe commented:
This is now ready to be reopened, and for you folks to take a look and let me know what you think

stephentoub commented on src/System.Runtime.Extensions/src/System/Net/WebUtility.cs:
```diff
@@ -357,6 +358,159 @@ private static byte[] UrlEncode(byte[] bytes, int offset, int count)
+                }
+                else if (!IsUrlSafeChar(s[i]))
+                {
+                    numUnsafe += GetByteCount(s, i, length);
```
Why did you implement GetByteCount manually rather than calling ```Encoding.UTF8.GetByteCount(s + i, length)```?

hughbe replied:
This implementation is considerably more efficient than Encoding.UtF8's implementation according to my benchmarks. Therefore, this custom implementation, and the one below are partly responsible for the increase in performance in cases where any kind of escaping is required

stephentoub replied:
I don't want us to have a separate implementation of UTF8 encoding.  If there are ways to improve the efficiency of UTF8Encoding, we should do so.

hughbe replied:
That's a fair shout. I just wish the code for UTF8Encoding was in corefx not coreclr so I could contribute

stephentoub replied:
Why can't you contribute to coreclr?

hughbe replied:
I'm under the impression they're not taking mscorlib contributions (at least until the release of 1.0)

stephentoub replied:
> I'm under the impression they're not taking mscorlib contributions (at least until the release of 1.0)

@jkotas?

davidsh replied:
We should not have our implementation for UTF8Encoding, even if it is "faster". We can live with the performance of the current implementation until such time as we are able to rev UTF8Encoding in CoreCLR.

hughbe replied:
That's fine. I'll store away the customised code for such a time that I'm able to fully understand the UTF8Encoding code!
The PR is still useful without the custom implementation: I'll update it shortly

stephentoub replied:
> We should not have our own implementation for UTF8Encoding, even if it is "faster"

Right, you and I are in agreement.

jkotas replied:
> I'm under the impression they're not taking mscorlib contributions (at least until the release of 1.0)

We can always discuss changes, and then decide whether and when it makes sense to take them.

I have written the current UTF8Encoding in mscorlib myself together with a few other people. It does number of perf tricks like processing multiple characters at a time - it should beat your impl on large strings. I believe that most of the overhead left in it is imposed by the surface area (abstract base class, customizable encoding fallbacks) - you can beat it on small strings because of this.

hughbe replied:
@jkotas I benchmarked my code vs yours for a long string - you're definitely right! Mine starts to perform worse almost exponentially. Therefore, its a good idea to keep using Encoding.UTF8 here

stephentoub commented on src/System.Runtime.Extensions/src/System/Net/WebUtility.cs:
```diff
@@ -357,6 +358,159 @@ private static byte[] UrlEncode(byte[] bytes, int offset, int count)
+                return new string(s, 0, length);
+            }
+
+            byte[] originalBytes = GetBytes(s, length);
```
Why did you implement GetBytes manually rather than allocating an array of the right length and using Encoding.UTF8.GetBytes that takes a char*?

hughbe replied:
See previous line note

hughbe commented:
I've updated the PR, removing my custom implementation of GetBytes. I had to keep part of GetByteCount because we are incrementing the byte count for each UrlUnafe char, not for each char

jkotas commented on src/System.Runtime.Extensions/src/System/Net/WebUtility.cs:
```diff
@@ -357,6 +358,93 @@ private static byte[] UrlEncode(byte[] bytes, int offset, int count)
+            if (numSpaces == 0 && numUnsafe == 0)
+            {
+                // Nothing to expand
+                return (string)originalString.Clone();
```
Just return the original string...

jamesqo commented on src/System.Runtime.Extensions/src/System/Net/WebUtility.cs:
```diff
@@ -357,6 +358,93 @@ private static byte[] UrlEncode(byte[] bytes, int offset, int count)
+            }
+        }
+
+        private static bool IsHighSurrogate(char c) => c >= HighSurrogateStart && c <= HighSurrogateEnd;
```
Could be replaced with `char.IsHighSurrogate` and `char.IsLowSurrogate`

jamesqo commented on src/System.Runtime.Extensions/src/System/Net/WebUtility.cs:
```diff
@@ -357,6 +358,93 @@ private static byte[] UrlEncode(byte[] bytes, int offset, int count)
+        private static bool IsHighSurrogate(char c) => c >= HighSurrogateStart && c <= HighSurrogateEnd;
+        private static bool IsLowSurrogate(char c) => c >= LowSurrogateStart && c <= LowSurrogateEnd;
+
+        private static unsafe int GetByteCount(char* pointer, int index, int pointerLength)
```
Just curious, why not use `Encoding.UTF8.GetByteCount` for this instead?

jamesqo commented on src/System.Runtime.Extensions/src/System/Net/WebUtility.cs:
```diff
@@ -357,6 +358,93 @@ private static byte[] UrlEncode(byte[] bytes, int offset, int count)
             return expandedBytes;
         }
 
+        private static unsafe string UrlEncode(string originalString, char* s, int length)
```
Nit: Maybe the name of the `char*` parameter should be changed to something like `pch`? I keep mistaking it for a string since it's named `s`.

jamesqo commented on src/System.Runtime.Extensions/src/System/Net/WebUtility.cs:
```diff
@@ -357,6 +358,93 @@ private static byte[] UrlEncode(byte[] bytes, int offset, int count)
+        {
+            int numSpaces = 0;
+            int numUnsafe = 0;
+            int numSafe = 0;
```
Maybe `numSpaces` and `numSafe` could be combined into one variable that keeps track of all the non-expanded characters?

hughbe replied:
Unfortunately not, as we want to keep track of the number of spaces, as spaces have to be escaped to '+'

jamesqo commented on src/System.Runtime.Extensions/src/System/Net/WebUtility.cs:
```diff
@@ -357,6 +358,93 @@ private static byte[] UrlEncode(byte[] bytes, int offset, int count)
+            fixed (byte* originalBytesPointer = originalBytes)
+            fixed (byte* newBytesPointer = newBytes)
+            {
+                Encoding.UTF8.GetBytes(s, length, originalBytesPointer, originalBytes.Length);
```
Note: we don't use the `char*` parameter past this point, so we're unnecessarily pinning the string while we're making a pass over `originalBytes`. You may want to refactor a bit to avoid that.

hughbe replied:
good point

hughbe commented:
@jamesqo thanks for the review, I hopefully addressed all your comments. I had to rebase the PR because git screwed up, but I also improved the perf of no escaping more by avoiding calling GetByteCount until we definitely know there's an invalid byte

jamesqo commented on src/System.Runtime.Extensions/src/System/Net/WebUtility.cs:
```diff
@@ -357,6 +357,66 @@ private static byte[] UrlEncode(byte[] bytes, int offset, int count)
+                    return originalString;
+                }
+                
+                int unsafeByteCount = Encoding.UTF8.GetByteCount(pch, originalString.Length) - safeCharCount - spaceCharCount;
```
@hughbe Thanks for replying to my review! What I actually had in mind when I said to use `GetByteCount` was per-character actually, something like this:

```cs
bool foundSpaces = false;
int unsafeCount = 0;

for (int i = 0; i < originalString.Length; i++)
{
    char c = pch[i];
    if (c == ' ') foundSpaces = true;
    else if (!IsUrlSafeChar(c))
    {
        if (i + 1 < originalString.Length && char.IsSurrogatePair(c, pch[i + 1]))
        {
            i++;
            unsafeCount += 4; // Surrogate pairs -> 4 chars
        }
        else
        {
            unsafeCount += Encoding.UTF8.GetByteCount(&c, 1);
        }
    }
}
```

The new implementation you have looks like it makes two passes over the string, one in the `for` loop and one in `GetByteCount`. (See [this](https://github.com/dotnet/corefx/pull/6698/files#diff-831332bbb389db07350457a4d92095f8R325) PR for what I did to handle this.)

hughbe replied:
Oh right - I don't have a PC on hand right now, but I seem to recall that this doesn't work for surrogate pairs because the length is 1? Is this right

jamesqo replied:
@hughbe That's why I added an `i++` statement to skip over the next low-surrogate char.

hughbe replied:
Understood,
I'm just wondering whether this is more or less performant (calculating the byte count for each char)
If you look above, @jkotas said that `Encoding.Utf8.GetByteCount` actually parses more than 1 char at a time, causing the performance to not suffer as much when strings get longer.
By calculating GetByteCount() for each unsafe char, are we forgoing that perf enhancement, and perhapss making things worse?
@jkotas do you have an opinion?

jkotas replied:
It depends on the ratio of safe and unsafe chars. I think it is better to do it for the whole string as it is right now.

hughbe commented:
@dotnet-bot Test Innerloop OSX Release Build and Test please

hughbe commented:
Test Innerloop OSX Release Build and Test please

jkotas commented on src/System.Runtime.Extensions/src/System/Net/WebUtility.cs:
```diff
@@ -357,6 +357,66 @@ private static byte[] UrlEncode(byte[] bytes, int offset, int count)
+                    Encoding.UTF8.GetBytes(pch, originalString.Length, originalBytesPointer, originalBytes.Length);
+
+                    int pos = 0;
+                    for (int i = 0; i < originalBytes.Length; i++)
```
This loop should be refactored into separate method and shared between `byte[]` and `string` handling codepaths.

jkotas commented on src/System.Runtime.Extensions/src/System/Net/WebUtility.cs:
```diff
@@ -357,6 +357,66 @@ private static byte[] UrlEncode(byte[] bytes, int offset, int count)
+                
+                int unsafeByteCount = Encoding.UTF8.GetByteCount(pch, originalString.Length) - safeCharCount - spaceCharCount;
+
+                byte[] originalBytes = new byte[safeCharCount + spaceCharCount + unsafeByteCount];
```
You can avoid this allocation. You can rewrite the bytes in place (make the output of `UTF8.GetBytes` to start at unsafeByteCount*2).

jkotas commented on src/System.Runtime.Extensions/src/System/Net/WebUtility.cs:
```diff
@@ -357,6 +357,66 @@ private static byte[] UrlEncode(byte[] bytes, int offset, int count)
+                    char c = pch[i];
+                    if (c == ' ')
+                    {
+                        spaceCharCount++;
```
This looks exactly like the loop in code that handles bytes. Make them look same (same variable names, formatting, ...).

jkotas commented:
There is way too much unsafe code than I would like to see. URL handling has a long history of buffer overruns - the buffer math is not super simple. I think this should be changed to just use safe code - large chunk of the performance improvement should still be there, even with safe code.

jkotas commented on src/System.Runtime.Extensions/src/System/Net/WebUtility.cs:
```diff
@@ -357,6 +357,66 @@ private static byte[] UrlEncode(byte[] bytes, int offset, int count)
             return expandedBytes;
         }
 
+        private static unsafe string UrlEncodeInternal(string originalString)
```
This should not be separate function since it is called just from one place and the caller has very little code it in. Fold it into the caller.

jkotas commented:
BTW: The IsUrlSafeChar method is likely a bottleneck - at least for the fast path. You can try replacing it with something like https://gist.github.com/jkotas/6a6337a606565517fd396bdfeab2462d and check whether it makes difference.

hughbe commented:
@jkotas thanks for your top advice and review - I agree with the unsafe stuff (which I'll remove entirely), and that extra removed allocation is great.

I won't pretend I fully understand the IsUrlSafeChar code, but I'll take a look at the perf impact ;)

hughbe commented:
I bencharked the IsUrlSafeChar method - seems like we're get a perf boost of up to 5-10%

davidsh commented:
I don't understand where we stand on this PR. Has all feedback been addressed?

hughbe commented:
@davidsh what hasn't been adressed

davidsh commented:
>what hasn't been addressed?

It's hard to tell from the conversation flow on the PR whether you acknowledged all the feedback and made changes. Some of the reviewers' comments didn't have comments from you indicating you made changes. It's too hard to just keep re-reviewing the sources and trying to map it back to comments.

@stephentoub Did you finish your review of this PR?

stephentoub commented:
> Did you finish your review of this PR?

No, I'd stopped once I saw that a custom UTF8 encoding implementation was being done.  I'll need to re-review.

hughbe commented:
Ah right, I'll remember to comment something like "done" next time
--> I've Defo addressed all of the comments made by @jkotas

hughbe commented:
@stephentoub that's been removed as it was considerably less performant for long strings

jkotas commented:
It looks good to me.

hughbe commented:
@jasonwilliams200OK sure, done. I had to rebase, as the branch was out of date and I couldn't restore packages locally to test

hughbe commented:
Test Innerloop OSX Debug Build and Test
Test  Innerloop OSX Release Build and Test

davidsh commented:
@stephentoub Do you have any more comments on this PR?

stephentoub commented:
Sorry for delay, I'll look shortly.

hughbe commented:
Test Innerloop OSX Debug Build and Test
Test  Innerloop OSX Release Build and Test

stephentoub commented on src/System.Runtime.Extensions/src/System/Net/WebUtility.cs:
```diff
@@ -331,30 +332,32 @@ private static byte[] UrlEncode(byte[] bytes, int offset, int count)
             // expand not 'safe' characters into %XX, spaces to +s
             byte[] expandedBytes = new byte[count + cUnsafe * 2];
-            int pos = 0;
+            GetEncodedBytes(bytes, 0, expandedBytes);
```
Is 0 correct here?  It doesn't need to be offset?

hughbe replied:
good catch. I'm surrprised there are no tests for this. I'll fix this and add a test

jkotas replied:
And it needs to take `count` argument too.

jkotas replied:
Could you please add a test case that would catch this bug?

stephentoub replied:
> And it needs to take  count  argument too.

Yup

> I'm surrprised there are no tests for this

Me, too :smile:

> I'll fix this and add a test

Thanks!

hughbe replied:
done

stephentoub commented on src/System.Runtime.Extensions/src/System/Net/WebUtility.cs:
```diff
@@ -367,9 +370,37 @@ public static string UrlEncode(string value)
+                else if (IsUrlSafeChar(ch))
+                {
+                    cSafe++;
+                }
```
Minor, but we probably expect url safe chars to be more common than spaces, right?  I'm wondering if the order of these checks should be reversed.  Doing so would match the order done earlier, and would mean that for url safe chars, we don't have to first check for whether it's a space.

hughbe replied:
done

stephentoub commented on src/System.Runtime.Extensions/src/System/Net/WebUtility.cs:
```diff
@@ -367,9 +370,37 @@ public static string UrlEncode(string value)
+                }
+            }
+
+            if (cSpaces == 0 && cSafe == value.Length)
```
Are both of these conditions necessary?  If cSafe == value.Length, how could cSpaces not be 0?

hughbe replied:
done

stephentoub commented on src/System.Runtime.Extensions/src/System/Net/WebUtility.cs:
```diff
@@ -367,9 +370,37 @@ public static string UrlEncode(string value)
+            // For perf, we avoid allocating a new array for the encoded bytes, and
+            // instead encode in place
+            byte[] newBytes = new byte[cSafe + cSpaces + unsafeByteCount + byteIndex];
+            Encoding.UTF8.GetBytes(value, 0, value.Length, newBytes, byteIndex);
```
This doesn't look right.  UTF8 can use 1-4 bytes per character.  newBytes has at most 3 bytes per character.  Is there going to be enough room for the worst case?  Plus, I don't understand why byteIndex is being passed in as the offset into newBytes.

hughbe replied:
We have tests encoding a surrogate pair, which pass (are those the worst case)

stephentoub replied:
Oh, wait, I see, unsafeByteCount already includes the UTF8 encoded count.

stephentoub commented on src/System.Runtime.Extensions/src/System/Net/WebUtility.cs:
```diff
@@ -367,9 +370,37 @@ public static string UrlEncode(string value)
+            byte[] newBytes = new byte[cSafe + cSpaces + unsafeByteCount + byteIndex];
+            Encoding.UTF8.GetBytes(value, 0, value.Length, newBytes, byteIndex);
+            
+            GetEncodedBytes(newBytes, byteIndex, newBytes);
```
I don't understand the logic here.  It seems like the goal is to use a single array.  But there's no guarantee that all of the characters that need URL encoding come at the end of the array... they could be interspersed throughout the entire array.

stephentoub commented on src/System.Runtime.Extensions/src/System/Net/WebUtility.cs:
```diff
@@ -331,30 +332,32 @@ private static byte[] UrlEncode(byte[] bytes, int offset, int count)
             // expand not 'safe' characters into %XX, spaces to +s
             byte[] expandedBytes = new byte[count + cUnsafe * 2];
-            int pos = 0;
+            GetEncodedBytes(bytes, offset, count,expandedBytes);
```
Nit: space before expandedBytes

stephentoub commented on src/System.Runtime.Extensions/src/System/Net/WebUtility.cs:
```diff
@@ -331,30 +332,32 @@ private static byte[] UrlEncode(byte[] bytes, int offset, int count)
+        private static void GetEncodedBytes(byte[] originalBytes, int startIndex, int count, byte[] newBytes)
+        {
+            int pos = 0;
+            for (int i = startIndex; i < startIndex + count; i++)
```
I'm assuming the JIT can see that startIndex and count will never change and can avoid redoing the addition on each iteration.  If that's not the case, you could put this upper bound into a local before the loop and then just use that.

stephentoub commented on src/System.Runtime.Extensions/src/System/Net/WebUtility.cs:
```diff
@@ -331,30 +332,32 @@ private static byte[] UrlEncode(byte[] bytes, int offset, int count)
+        }
 
-            for (int i = 0; i < count; i++)
+        private static void GetEncodedBytes(byte[] originalBytes, int startIndex, int count, byte[] newBytes)
```
Nit: can we use the name offset instead of startIndex?  That'll help to be consistent with the call sites and typical usage in APIs like this.

stephentoub commented on src/System.Runtime.Extensions/src/System/Net/WebUtility.cs:
```diff
@@ -331,30 +332,32 @@ private static byte[] UrlEncode(byte[] bytes, int offset, int count)
-            for (int i = 0; i < count; i++)
+        private static void GetEncodedBytes(byte[] originalBytes, int startIndex, int count, byte[] newBytes)
+        {
+            int pos = 0;
```
For the case where originalBytes and newBytes are the same array, could you add some debug asserts to verify that we never get to the point of overwriting input from originalBytes that hasn't yet been read as we write to newBytes?

hughbe replied:
done

stephentoub commented on src/System.Runtime.Extensions/src/System/Net/WebUtility.cs:
```diff
@@ -367,9 +370,37 @@ public static string UrlEncode(string value)
+            int byteIndex = unsafeByteCount * 2;
+            
+            // For perf, we avoid allocating a new array for the encoded bytes, and
+            // instead encode in place
```
I think it'd be worth expanding on this comment, e.g.
"Instead of allocating one array of length ```Encoding.UTF8.GetByteCount(value)``` to store the UTF8 encoded bytes and then a second array of length ```cSafe + cSpaces + 3 * (Encoding.UTF8.GetByteCount(value) - cSafe - cSpaces)``` to store the URL encoded UTF8 bytes, we allocate a single array of length ```cSafe + cSpaces + Encoding.UTF8.GetByteCount(value) * 3```, saving ```Encoding.UTF8.GetByteCount(value) - 3*(cSafe +cSpaces)``` bytes allocated.  We encode the UTF8 to the end of this array, and then URL encode to the beginning of the array."
or something like that.  If you decide to include a comment like this, including the math, please double-check my math :smile:

hughbe replied:
I'll do this before you merge - I just realised I didn't push a file change so the thing doesn't compile right now

hughbe replied:
done - the math looks correct to me!

stephentoub commented:
Thanks for the perf speedups, @hughbe. LGTM.

davidsh commented:
@hughbe Please squash these commits.  Then we'll merge when green.  Thx.

stephentoub commented:
@davidsh, we can actually now squash as part of merging.  Relatively new GitHub feature we've been using.

jkotas commented:
One of the commits with my contribution is intentionally factored out...

stephentoub commented:
> One of the commits with my contribution is intentionally factored out...

Ah, ok.  Then yeah, the others should be squashed.

hughbe commented:
I'll take a look - I might have to reorder the commits or something like that, hopefully git will allow me to

stephentoub commented:
> I'll take a look - I might have to reorder the commits or something like that, hopefully git will allow me to

If it's a problem, you can squash to those before Jan's commit and then have a separate commit of addressed feedback after his commit.  Doesn't need to be perfect.

hughbe commented:
It was pretty easy - `git rebase -i HEAD~N` lets you reorder commits

davidsh commented:
Test  Innerloop OSX Release Build and Test

davidsh commented:
@dotnet-bot  Innerloop OSX Release Build and Test

danmosemsft commented:
@hughbe I wonder if HttpUtility.UrlEncode (which is implemented in HttpEncoder) could use comparable optimization

hughbe commented:
@danmosemsft I'm surprised they don't share code (any reason for that?)

I'll examine in the weekend

danmosemsft commented:
I think S.Net.WebUtility was cloned from S.Web.HttpUtility originally because the latter was on System.Web.dll and folks didn't want to pull that in just to encode a URL.

hughbe commented:
What about extracting common code to a src/common - that would avoid the duplication and give us the Perf improvements for both

danmosemsft commented:
If that's feasible it sounds good to me!
